lxd/qmp/monitor.go

```diff
@@ -0,0 +1,276 @@
+package qmp
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/digitalocean/go-qemu/qmp"
+
+	"github.com/lxc/lxd/shared"
+)
+
+var monitors = map[string]*Monitor{}
+var monitorsLock sync.Mutex
+
+// Monitor represents a QMP monitor.
+type Monitor struct {
+	path string
+	qmp  *qmp.SocketMonitor
+
+	agentReady   bool
+	disconnected bool
+	chDisconnect chan struct{}
+	eventHandler func(name string, data map[string]interface{})
+}
+
+// Connect creates or retrieves an existing QMP monitor for the path.
+func Connect(path string, eventHandler func(name string, data map[string]interface{})) (*Monitor, error) {
+	monitorsLock.Lock()
+	defer monitorsLock.Unlock()
+
+	// Look for an existing monitor.
+	monitor, ok := monitors[path]
+	if ok {
+		monitor.eventHandler = eventHandler
+		return monitor, nil
+	}
+
+	// Setup the connection.
+	qmpConn, err := qmp.NewSocketMonitor("unix", path, time.Second)
+	if err != nil {
+		return nil, err
+	}
+
+	err = qmpConn.Connect()
+	if err != nil {
+		return nil, err
+	}
+
+	// Setup the monitor struct.
+	monitor = &Monitor{}
+	monitor.path = path
+	monitor.qmp = qmpConn
+	monitor.chDisconnect = make(chan struct{}, 1)
+	monitor.eventHandler = eventHandler
+
+	// Spawn goroutines.
+	err = monitor.run()
+	if err != nil {
+		return nil, err
+	}
+
+	// Register in global map.
+	monitors[path] = monitor
+
+	return monitor, nil
+}
+
+func (m *Monitor) run() error {
+	// Start ringbuffer monitoring go routine.
+	go func() {
+		for {
+			// Read the ringbuffer.
+			resp, err := m.qmp.Run([]byte(`{"execute": "ringbuf-read", "arguments": {"device": "vserial", "size": 16, "format": "utf8"}}`))
```

Comment:
@stgraber is 16 bytes the max length of any of our ring-buffer messages?

Reply:
Yes, the serial device is configured with a size of 16B.

Reply:
@stgraber thanks, I wonder if its worth defining a constant for that value and then using it here and down on line https://github.com/lxc/lxd/pull/6527/files#diff-8c876c3968dd86efda1f7fcd87e6c34aR1027 to show they are related and avoid any bugs where one is increased without the other.